File: 1046-last-stone-weight/1046-last-stone-weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        
        maxHeap = []
        
        for num in stones:
            heapq.heappush(maxHeap, -1 * num)
        
        while len(maxHeap) > 1:
            biggest = -1 * heapq.heappop(maxHeap)
            smallest = -1 * heapq.heappop(maxHeap)
            
            diff = biggest - smallest
            if diff != 0:
                heapq.heappush(maxHeap, -1 * diff)
        
        if len(maxHeap) == 0: return 0
        else: return -1 * maxHeap[0]
        

File: 1046-last-stone-weight/1046-last-stone-weight.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int, vector<int>> pq;
        for (auto stone: stones) {
            pq.push(stone);
        }
        
        while(pq.size() > 1) {
            int stone1 = pq.top();
            pq.pop();
            int stone2 = pq.top();
            pq.pop();
            int val = stone1 - stone2;
            if (val != 0) {
                pq.push(val);
            }
        }
        
        return pq.size() == 0 ? 0 : pq.top();
    }
};